fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogBottomSheet.kt

```diff
@@ -127,19 +135,29 @@ fun TranslationsDialogBottomSheet(
 
         Spacer(modifier = Modifier.height(14.dp))
 
-        if (translationError !is TranslationError.CouldNotLoadLanguagesError) {
+        if (translationError !is TranslationError.CouldNotLoadLanguagesError &&
+            translateFromLanguages != null && translateToLanguages != null
```

Comment:
I'm wondering whether we should introduce a new `Error` type that can be used in `GeckoTranslateSessionDelegate::onTranslationStateChange` to avoid some of this nullability. Are there cases where the first part of this conditional is true and the latter parts are false?

Reply:
Or perhaps we need `TranslationEngineState` to be an `Option` type? Like:

```
sealed class TranslationEngineState {
    data class Success(val candidateLanguages, from, to, etc) : TranslationEngineState
    data class Error(val error)  
}
```

Reply:
> translateFromLanguages != null && translateToLanguages != null

I think the only scenario this will happen is if fetching the languages is late, but there is no error yet, then this can happen. 

> initialFrom and intitialTo

Very often we can have one and not the other:
* `intitialFrom` - we detected a compatible page language
   * null if we can't detect page language
   * null if page language is unsupported (have another [error case ](https://www.figma.com/file/GT6oZLFw32LiCwKl5GOMWD/Full-Page-Translation?type=design&node-id=7202-119890&mode=design&t=qr7VxyeKAmJhSnfM-0)that'll need to be integrated) 
* `initialTo` - user's preferred language
   * null if user's preferred language isn't supported
   * maybe can also be null for other reasons, like we can't determine it or the language code doesn't parse


> I'm wondering whether we should introduce a new Error type that can be used in GeckoTranslateSessionDelegate::onTranslationStateChange to avoid some of this nullability.

I could see sending an action that does something from `onTranslationStateChange`.

Mentioning directly some background here, because the naming was really tricky on that state and a lot happens there. The `TranslationEngineState` is fully controlled by the browser engine independently of AC and Fenix and we are just reporting back details.

`TranslationEngineState` generally changes many times:
* Before page load (I think, very early at any point) - lots of expected nulls
* After page load - some expected nulls
* Translate - almost everything full
* Translate Restore - looks like after page load
* Restoring page from a full shutdown

The [GeckoView code](https://searchfox.org/mozilla-central/rev/24f9846b7d8ea0819edee7aafc51a39753f1619a/mobile/android/geckoview/src/main/java/org/mozilla/geckoview/TranslationsController.java#1173,1184) kinda shows the relationships happening in that delegate nicely. The state comes in as a [message from toolkit](https://searchfox.org/mozilla-central/rev/24f9846b7d8ea0819edee7aafc51a39753f1619a/mobile/android/modules/geckoview/GeckoViewTranslations.sys.mjs#27,36,106-123) that performs the translation. 

One reason it is especially important is this translations engine can make decisions independent of Fenix based on some prefs we set. For example, I can set always translate French and right after page load, the page will translate without me touching Fenix UI. Fenix UI needs to reflect this browser state; however.

Other tangent:

I don't think we need that specific [TranslationEngineState.error](https://searchfox.org/mozilla-mobile/rev/99576d6e69eac8ee1bf6c8a654d5fa96a6f6bf84/firefox-android/android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/TranslationEngineState.kt#11) the engine reports except for debugging. It is the error the browser reported and, IIRC, the error Desktop will visually show. `translationError` [here](https://searchfox.org/mozilla-mobile/rev/99576d6e69eac8ee1bf6c8a654d5fa96a6f6bf84/firefox-android/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/TranslationsState.kt#33) is the one important to Fenix.

>Or perhaps we need TranslationEngineState to be an Option type? 

Sure, let me know what you prefer based on the addition info! I don't think I have enough background to evaluate what should be done here.

Reply:
Thanks for all the info! The `Option` type isn't seeming worth it based on all of that. Mostly just asking questions out loud to try and understand the domain better

Reply:
Thanks for asking questions and brainstorming. It is helping me understand my own assumptions better for sure!